src/wristpy/processing/analytics.py

```diff
@@ -0,0 +1,305 @@
+
+        return sleep_onset_wakeup
+
+    def _spt_window(
+        self, anglez_data: models.Measurement, threshold: float = 0.13
+    ) -> models.Measurement:
+        """Implement Heuristic distribution of z angle (HDCZA) for SPT window detection.
+
+        This function finds the absolute difference of the anglez data over 5s windows.
+        We find when the 5-minute rolling median of that difference is below a
+        threshold (default set to minimum in GGIR range).
+        We then find blocks of 30 minutes (360 blocks of 5s) where the rolling median is
+        below the threshold.
+        Any gaps in SPT windows that are less than a specified window length
+        (default 60 minutes) are filled.
+
+        Args:
+            anglez_data: the raw anglez data, calculated from calibrated acceleration.
+            threshold: the threshold for the distribution of z angle.
+
+        Returns:
+            A Measurement instance with values set to 1 indicating identified
+            SPT windows and corresponding time stamps.
+        """
+        anglez_abs_diff = self._compute_abs_diff_mean_anglez(anglez_data)
+        anglez_median_long_epoch = computations.moving_median(anglez_abs_diff, 300)
+        below_threshold = (
+            (anglez_median_long_epoch.measurements < threshold).astype(int).flatten()
+        )
+
+        sleep_idx_array = self._find_long_blocks(below_threshold)
+
+        sleep_idx_array_filled = self._fill_short_blocks(sleep_idx_array)
+
+        return models.Measurement(
+            measurements=sleep_idx_array_filled, time=anglez_median_long_epoch.time
+        )
+
+    def _calculate_sib_periods(
+        self, anglez_data: models.Measurement, threshold_degrees: int = 5
+    ) -> models.Measurement:
+        """Find the sustained inactivity bouts.
+
+        This function finds the absolute difference of the anglez data over 5s windows.
+        We then find the 5-minute windows where all of the differences are below a
+        threshold (defaults to 5 degrees).
+
+        Args:
+            anglez_data: the raw anglez data, calculated from calibrated acceleration.
+            threshold_degrees: the threshold, in degrees, for inactivity.
+
+        Returns:
+            A Measurement instance with values set to 1 indicating identified SIB
+            windows, and corresponding time stamps.
+        """
+        anglez_abs_diff = self._compute_abs_diff_mean_anglez(anglez_data)
+
+        anglez_pl_df = pl.DataFrame(
+            {"time": anglez_abs_diff.time, "angz_diff": anglez_abs_diff.measurements}
+        )
+
+        anglez_pl_df = anglez_pl_df.with_columns(pl.col("time").set_sorted())
+        anglez_grouped_by_window_length = anglez_pl_df.group_by_dynamic(
+            index_column="time", every="5m"
+        ).agg([pl.all().exclude(["time"])])
+        flag = anglez_grouped_by_window_length["angz_diff"].map_elements(
+            lambda lst: all(x < threshold_degrees for x in lst)
```

Comment:
is the loop necessary, wouldn't checking the max work ?

Reply:
So `anglez_grouped_by_window_length["angz_diff"]` is a pl.Series where each element is a list. This loop checks whether all data points that makes up each list (I improved the naming convention of `x` and `lst` to hopefully clarify this) is below the threshold or not

FYI it looks like this now:
```
   flag = anglez_grouped_by_window_length["angz_diff"].map_elements(
            lambda grouped_list_data: all(
                list_data < threshold_degrees for list_data in grouped_list_data
            )
        )
 ```